google/cloud/internal/tracing_rest_client.cc

```diff
@@ -0,0 +1,114 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/internal/tracing_rest_client.h"
+#include "google/cloud/internal/opentelemetry.h"
+#include "google/cloud/internal/rest_opentelemetry.h"
+#include "google/cloud/internal/tracing_http_payload.h"
+#include "google/cloud/internal/tracing_rest_response.h"
+
+namespace google {
+namespace cloud {
+namespace rest_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+void ExtractAttributes(StatusOr<std::unique_ptr<RestResponse>> const& value,
+                       opentelemetry::trace::Span& span) {
+  if (!value) return;
+  auto const& response = *value;
+  if (!response) return;
+  for (auto const& kv : response->Headers()) {
+    span.SetAttribute("http.response.header." + kv.first, kv.second);
+  }
+}
+
+/**
+ * Extracts information from @p value, and adds it to a span.
+ *
+ * The span is ended. The original value is returned, for the sake of
+ * composition.
+ *
+ * Note that this function should be called after the RPC has finished.
+ */
+StatusOr<std::unique_ptr<RestResponse>> EndSpan(
+    opentelemetry::trace::Span& request_span,
+    StatusOr<std::unique_ptr<RestResponse>> value) {
+  if (!value) return internal::EndSpan(request_span, std::move(value));
+  auto payload_span = MakeSpanHttpPayload(request_span);
+  ExtractAttributes(value, request_span);
+  value = std::make_unique<TracingRestResponse>(*std::move(value),
+                                                std::move(payload_span));
+  return internal::EndSpan(request_span, std::move(value));
+}
+
+TracingRestClient::TracingRestClient(std::unique_ptr<RestClient> impl)
+    : impl_(std::move(impl)) {}
+
+StatusOr<std::unique_ptr<RestResponse>> TracingRestClient::Delete(
+    RestRequest const& request) {
+  auto span = MakeSpanHttp(request, "DELETE");
```

Comment:
FYI: We still need to do context propagation for HTTP. So in a future PR, we should be setting the span as active, and injecting the trace context as a header.

Basically this class, but setting metadata on a RestRequest

https://github.com/googleapis/google-cloud-cpp/blob/f5f98d27ec1ebb23edc5249e67421fa5fb49159c/google/cloud/internal/grpc_opentelemetry.cc#L36-L66

https://github.com/googleapis/google-cloud-cpp/blob/f5f98d27ec1ebb23edc5249e67421fa5fb49159c/generator/integration_tests/golden/v1/internal/golden_kitchen_sink_tracing_stub.cc#L49